src/org/labkey/snd/query/AttributeDataTable.java

```diff
@@ -402,31 +376,32 @@ else if (stringValue != null)
         public int mergeRows(User user, Container container, DataIteratorBuilder rows, BatchValidationException errors,
                              @Nullable Map<Enum, Object> configParameters, Map<String, Object> extraScriptContext)
         {
+            Logger log = SNDManager.getLogger(configParameters, AttributeDataTable.class);
+
+            // Large merge triggers importRows path
+            if (getRowCount(rows, configParameters, errors) > SNDManager.MAX_MERGE_ROWS)
+            {
+                log.info("More than " + SNDManager.MAX_MERGE_ROWS + " rows. using importRows method.");
+                return importRows(user, container, rows, errors, configParameters, extraScriptContext);
+            }
+            log.info("Merging rows.");
             List<Map<String, Object>> data = getMutableData(rows, getDataIteratorContext(errors, InsertOption.MERGE, configParameters));
```

Comment:
You're essentially streaming all the data twice here.  First with getRowCount, which uses getMutableData, and then again with getMutableData.  I would consider just collecting the data once using getMutableData, then looking at the size of the List to see if its greater than MAX_MERGE_ROWS

Reply:
Yep - easy fix. I wish there was a better way to get the row count.